src/main/java/week2/BTree.java

```diff
@@ -0,0 +1,113 @@
+package week2;
+
+
+class BTree<T extends Comparable<T>> {
+
+  private boolean bothFound = false;
+  private Node<T> root;
+
+  BTree(Node<T> root) {
+    this.root = root;
+  }
+
+
+  /* Question 1: Given a Binary Tree and a key, write a function that prints all the ancestors
+                 of the key in the given binary tree.
+                 Example: For the key 5 and the following tree we should print: 3, 9, 16.
+
+     Overview: We define a private method to recursively parse through the tree searching for the
+              element. Once the value is found, the method returns true to signal up to call stack
+               that element is found.
+
+     Time: O(N) where N is the number of elements in tree
+
+   */
+  String printAncestors(T key) {
+    StringBuilder sb = new StringBuilder();
+    printAncestors(key, root, sb);
+
+    return sb.toString();
+  }
+
+  private boolean printAncestors(T key, Node<T> element, StringBuilder sb) {
+    if (element != null) {
+      int comparision = element.getKey().compareTo(key);
+
+      if (comparision == 0) {
+        //signal back that element is found and hence the recursion path back are the ancestors
+        return true;
+      } else if (printAncestors(key, element.getLeft(), sb) || printAncestors(key,
+          element.getRight(), sb)) {
+        sb.append(sb.length() > 0 ? ", " : "");
+        sb.append(element.getKey().toString());
+
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /* Question 2: Design an algorithm and write code to find the lowest common ancestor of two nodes
+                 in a binary tree. Avoid storing additional nodes in a data structure
+
+     Overview: The solution is implemented recursively.
+
+               Possible alternatives (omitted duals) considered in the recursive commonAncestor method:
+               1) None of the element are in the tree (else branch in commonAncestor)
+               2) One of the element is in the tree (bothFound is not set to true)
+               3) key1 is an ancestor of key2 => key2 (else if branch)
+               4) key1 is descendant on the left of some node and key2 is a descendant on thr right of the same node (if branch)
+
+      Time: O(N) where N is the number of elements in tree
+   */
+  T commonAncestor(T key1, T key2) {
+    bothFound = false;
+    T res = commonAncestor(key1, key2, root);
+
+    return bothFound ? res : null;
+  }
+
+  private T commonAncestor(T key1, T key2, Node<T> element) {
```

Comment:
I'm not a great fan of the way that this code assumes that each key only appears once in the tree, so that if the same key appears twice, you'll find the common ancestor of that key. It's a bit brittle. I can't really say it's categorically wrong, though, if we are guaranteed nodes don't repeat!

You return a T, and also update the bothFound variable. I think what you'd like to do is return a pair of the T value and the "bothFound" value, but Java doesn't really make that easy. You can return an instance of a small helper class containing the two values, but it's a bit grim.

I also find your use of the return value and the boolean flag a little confused. Note that when you're finding a value in the children, you're returning the value, just to say you've found it - you never actually use it (beyond checking it's not null). Only when you find the ancestor node do you pass that up the tree, having set "bothFound".

So, a cleaner way to write this, if you keep the return value and the extra variable, is to have the method return a boolean for "found one of the things we were looking for", and the extra variable is the ancestor, which is set to the current node when either both children or a child and the current node are elements being looked for.